### PR TITLE
SOTW 2 script

### PR DESCRIPTION
**Pull request by htroisi**

### CODE REVIEW

**Review by zasgar — CHANGES_REQUESTED** [review]
I would call it sotw so it's not as long of a name. 


At a high level, I think we need to take an argument for the domain name and also parse out the requests to make it a bit easier to see in the table.

**Comment by htroisi** [comment]
Requests are parsed to see query + answers:
Screen Shot 2020-11-11 at 9 26 26 AM

File: pxl_scripts/sotw/ndots/data.pxl [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+# Prep
+# kubectl apply -f https://k8s.io/examples/admin/dns/dnsutils.yaml
+
+# Then trigger a DNS request
+# kubectl exec -i -t dnsutils -- nslookup news.ycombinator.com
+
+# ----------------------------------------------------------------
+# Script variables
+# ----------------------------------------------------------------
+max_num_records = 1000
```
Comment (oazizi000) [inline pxl_scripts/sotw/ndots/data.pxl:18]: We should probably pull out variables like this to be script argument.

In this case, we can also just completely remove this and the corresponding `df.head` below.

File: pxl_scripts/sotw/ndots/data.pxl [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+''' Kubernetes ndots demo
+'''
+
+import px
+
+# Prep
+# kubectl apply -f https://k8s.io/examples/admin/dns/dnsutils.yaml
+
+# Then trigger a DNS request
+# kubectl exec -i -t dnsutils -- nslookup news.ycombinator.com
```
Comment (oazizi000) [inline pxl_scripts/sotw/ndots/data.pxl:13]: nit: Change to:

```
# Trigger a DNS request by deploying a dnsutils pod, and issuing a DNS request.
# kubectl apply -f https://k8s.io/examples/admin/dns/dnsutils.yaml
# kubectl exec -i -t dnsutils -- nslookup news.ycombinator.com
```

**Review by oazizi000 — APPROVED** [review]

File: pxl_scripts/sotw/dns_external_fqdn_list/data.pxl [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    # The dns_events table pairs DNS requests with their responses.
+    df = px.DataFrame(table='dns_events', start_time=start)
+
+    # Parse the DNS respoonse to determine if it was successfully resolved.
+    df.resp_body = px.pluck(df.resp_body, 'answers')
+    df.request_resolved = px.contains(df.resp_body, 'name')
+
+    # Parse the DNS request for query name.
+    df.req_body = px.pluck(df.req_body, 'queries')
+    df.idx1 = px.find(df.req_body, '\"name\":')
```
Comment (nserrino) [inline pxl_scripts/sotw/dns_external_fqdn_list/data.pxl:21]: add note that this logic will be cleaned up when we support object types

File: pxl_scripts/sotw/dns_queries_filtered/data.pxl [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	# Convert DNS IP to string.
+	df.dns_server = px.Service(px.nslookup(df.remote_addr))
+
+	# Parse the DNS respoonse to determine if it was successfully resolved.
+	df.resp_body = px.pluck(df.resp_body, 'answers')
+	df.request_resolved = px.contains(df.resp_body, 'name')
+
+	# Parse the DNS request for query name.
+	df.req_body = px.pluck(df.req_body, 'queries')
+	df.idx1 = px.find(df.req_body, '\"name\":')
```
Comment (nserrino) [inline pxl_scripts/sotw/dns_queries_filtered/data.pxl:27]: similar, note here that this logic will get cleaner when we add support for object types

File: pxl_scripts/sotw/dns_queries_filtered/vis.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{
+  "variables": [
+    {
+      "name": "start",
+      "type": "PX_STRING",
+      "description": "The start time of the window in time units before now.",
+      "defaultValue": "-3h"
```
Comment (nserrino) [inline pxl_scripts/sotw/dns_queries_filtered/vis.json:7]: we typically do a -5m start time, we might want to be consistent

**Review by nserrino — APPROVED** [review]

File: pxl_scripts/sotw/dns_external_fqdn_list/data.pxl [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    # Parse the DNS request for query name.
+    # TODO: cleanup this logic when we support object types
+    df.req_body = px.pluck(df.req_body, 'queries')
+    df.idx1 = px.find(df.req_body, '\"name\":')
+    df.fqdn_partial = px.substring(df.req_body, df.idx1 + 8, px.length(df.req_body) - (df.idx1 + 8) - 3)
+    df.idx2 = px.find(df.fqdn_partial, ',')
+    df.fqdn = px.substring(df.fqdn_partial, 0, df.idx2 - 1)
+
+    # Filter for resolved requests.
+    df = df[df.request_resolved == True]
```
Comment (oazizi000) [inline pxl_scripts/sotw/dns_external_fqdn_list/data.pxl:28]: minor: What do you think about moving this up to immediately after df.request_resolved is computed?

File: pxl_scripts/sotw/dns_queries_filtered/vis.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      "name": "start",
+      "type": "PX_STRING",
+      "description": "The start time of the window in time units before now.",
+      "defaultValue": "-5m"
+    },
+    {
+      "name": "query_name_filter",
+      "type": "PX_STRING",
+      "description": "The partial query name used to filter the DNS requests by.",
+      "defaultValue": "news.ycombinator.com"
```
Comment (oazizi000) [inline pxl_scripts/sotw/dns_queries_filtered/vis.json:13]: Probably should have a different placeholder.

File: pxl_scripts/sotw/dns_external_fqdn_list/data.pxl [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+'''
+
+import px
+
+def fqdn_list(start: str):
+
+    # The dns_events table pairs DNS requests with their responses.
+    df = px.DataFrame(table='dns_events', start_time=start)
+
+    # Parse the DNS respoonse to determine if it was successfully resolved.
```
Comment (oazizi000) [inline pxl_scripts/sotw/dns_external_fqdn_list/data.pxl:15]: typo: response

File: pxl_scripts/sotw/dns_queries_filtered/data.pxl [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+''' DNS Queries filtered by query name. 
+Lists all DNS queries filtered by a specific query name. 
+'''
+
+import px
+
+def dns_requests(start: str, query_name_filter: str):
+
+    # The dns_events table pairs DNS requests with their responses.
+	df = px.DataFrame(table='dns_events', start_time=start)
```
Comment (oazizi000) [inline pxl_scripts/sotw/dns_queries_filtered/data.pxl:13]: nit: indentation

File: pxl_scripts/sotw/dns_queries_filtered/data.pxl [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	# Parse the DNS request for query name.
+	# TODO: cleanup this logic when we support object types
+	df.req_body = px.pluck(df.req_body, 'queries')
+	df.idx1 = px.find(df.req_body, '\"name\":')
+	df.query_name_partial = px.substring(df.req_body, df.idx1 + 8, px.length(df.req_body) - (df.idx1 + 8) - 3)
+	df.idx2 = px.find(df.query_name_partial, ',')
+	df.dns_query_name = px.substring(df.query_name_partial, 0, df.idx2 - 1)
+	df = df.drop(['idx1', 'idx2'])  
+
+	## Step 2: All pod / dns_server pairings containing FQDN as substring.
```
Comment (oazizi000) [inline pxl_scripts/sotw/dns_queries_filtered/data.pxl:34]: nit: Step 2 without a Step 1.

Comment (oazizi000) [inline pxl_scripts/sotw/dns_queries_filtered/data.pxl:34]: Also could we make "containing FQDN as substring" more clear. I think we're trying to say "containing search substring" or "containing query_name_filter".

File: pxl_scripts/sotw/dns_external_fqdn_list/data.pxl [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    df.fqdn_partial = px.substring(df.req_body, df.idx1 + 8, px.length(df.req_body) - (df.idx1 + 8) - 3)
+    df.idx2 = px.find(df.fqdn_partial, ',')
+    df.fqdn = px.substring(df.fqdn_partial, 0, df.idx2 - 1)
+
+    # Filter for resolved requests.
+    df = df[df.request_resolved == True]
+
+    # Filter out queries for local domain names.
+    df = df[not px.contains(df.fqdn,'.local')]
+    df = df[not px.contains(df.fqdn,'.internal')]
```
Comment (oazizi000) [inline pxl_scripts/sotw/dns_external_fqdn_list/data.pxl:32]: Worried this is too specific to our k8s setup. I believe the suffix can change based on the k8s config.

Not sure there's a quick workaround though, so we might just have to document this better.

Comment (htroisi) [reply pxl_scripts/sotw/dns_external_fqdn_list/data.pxl:32]: I think we're going to have to go with better documentation here. I couldn't come up with a better way to identify external domains.

File: pxl_scripts/sotw/dns_external_fqdn_list/data.pxl [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+    # Parse the DNS request for query name.
+    # TODO: cleanup this logic when we support object types.
+    df.req_body = px.pluck(df.req_body, 'queries')
+    df.idx1 = px.find(df.req_body, '\"name\":')
+    df.fqdn_partial = px.substring(df.req_body, df.idx1 + 8, px.length(df.req_body) - (df.idx1 + 8) - 3)
+    df.idx2 = px.find(df.fqdn_partial, ',')
+    df.fqdn = px.substring(df.fqdn_partial, 0, df.idx2 - 1)
+
+    # Filter out local domains quieres by excluding fqdns with known local domain suffix.
```
Comment (oazizi000) [inline pxl_scripts/sotw/dns_external_fqdn_list/data.pxl:30]: How about an additional comment like "If your k8s cluster uses a different internal domain suffix, add it here."

**Review by oazizi000 — APPROVED** [review]

**Review by zasgar — APPROVED** [review]

**Pull request merged**